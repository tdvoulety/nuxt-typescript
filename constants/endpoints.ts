export const COMPANIES_URL: string = '/companies'
export const SEARCH_APP_URL: string = '/search_app/api/v1/'
export const BOOKMARKS_URL: string = '/bookmarks/api/v1/company_bookmarks'
export const TRACKING_EVENTS_URL: string = '/octopus/api/events'
export const TRACKING_PAGES_URL: string = '/octopus/api/pages'
export const SWITCHER_URL: string = '/internal_api/v1/flags?only='
