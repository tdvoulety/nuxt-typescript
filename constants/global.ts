interface Component {
  key: string,
  component: object
}

export const GLOBAL_COMPONENTS: Component[] = []

export const GLOBAL_CLICK_EVENT: string = 'GLOBAL_CLICK_EVENT'
export const GLOBAL_RESIZE_EVENT: string = 'GLOBAL_RESIZE_EVENT'
