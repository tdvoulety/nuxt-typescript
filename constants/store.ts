export const NAMESPACES = {
  I18N: 'i18n/',
  AUTH: 'auth/',
}

export const ACTIONS = {
  NUXT_SERVER_INIT: 'nuxtServerInit',
  INIT: 'INIT',
  GET_ITEMS: 'GET_ITEMS',
  GET_DETAIL: 'GET_DETAIL',
}

export const MUTATIONS = {
  SET_BASE: 'SET_BASE',
  SET_LOCALE: 'SET_LOCALE',
  SET_SIDE_NAV_VISIBILITY: 'SET_SIDE_NAV_VISIBILITY',
}
