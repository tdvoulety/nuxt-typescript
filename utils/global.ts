// @ts-nocheck
export const isInteger = (value) => typeof value === 'number' && Number.isFinite(value) && Math.floor(value) === value

export const isPromise = (value) => Boolean(value && typeof value.then === 'function')

export const isFunction = (value) => Boolean(value && typeof value === 'function')

export const isNumeric = (value) => !Number.isNaN(parseFloat(value)) && Number.isFinite(value)

export const cloneDeep = (value) => {
  if (value === null || value === undefined) {
    return value
  }

  return JSON.parse(JSON.stringify(value))
}

export const elementIsVisible = (element) => element && element.offsetWidth > 0 && element.offsetHeight > 0
