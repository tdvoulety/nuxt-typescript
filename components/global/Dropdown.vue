<template>
  <div class="dropdown" :class="{ 'show': opened, 'disabled': disabled }">
    <button type="button" class="dropdown__toggle" :id="dropdownId" :disabled="disabled">
      <span :id="dropdownTextId">{{ text }}</span>
    </button>

    <ul role="menu" tabindex="-1" class="dropdown__menu" :class="{ 'show': opened }" @click.stop>
      <li v-for="item in items" :key="item.value">
        <button type="button" class="dropdown__item" @click="handleClickOnItem(item)">
          {{ item.text }}
        </button>
      </li>
    </ul>
  </div>
</template>

<script>
import { GLOBAL_CLICK_EVENT } from '@/constants/global'

export default {
  props: {
    text: {
      type: [String, Number],
      required: true,
      default: '',
    },
    items: {
      type: Array,
      required: true,
      default: () => [],
    },
    disabled: {
      type: Boolean,
      required: false,
      default: false,
    },
  },
  data() {
    return {
      opened: false,
      id: null,
    }
  },
  computed: {
    dropdownId() {
      return `dropdown-${this.id}`
    },
    dropdownTextId() {
      return `${this.dropdownId}__text`
    },
  },
  methods: {
    toggleMenu() {
      this.opened = !this.opened
    },
    handleGlobalClickEvent(e) {
      if (!this.disabled && e && (e.target.id === this.dropdownId || e.target.id === this.dropdownTextId)) {
        this.toggleMenu()
        e.stopPropagation()
      } else {
        this.hideMenu()
      }
    },
    hideMenu() {
      this.opened = false
    },
    handleClickOnItem(item) {
      this.hideMenu()
      this.$emit('onClick', item)
    },
  },
  mounted() {
    // eslint-disable-next-line no-underscore-dangle
    this.id = this._uid
    this.$root.$on(GLOBAL_CLICK_EVENT, this.handleGlobalClickEvent)
  },
  destroyed() {
    this.$root.$off(GLOBAL_CLICK_EVENT, this.handleGlobalClickEvent)
  },
}
</script>

<style scoped lang="scss">
.dropdown {
  position: relative;
  vertical-align: middle;

  &__toggle {
    font-family: 'Source Sans Pro', sans-serif;
    font-size: 1.4rem;
    white-space: nowrap;
    outline: none;
    cursor: pointer;
    margin: 0;
    padding: 0;
    border: 0;
    background: 0;
    color: inherit;

    &:hover, &:focus {
      outline: none;
    }

    &:after {
      display: inline-block;
      margin-left: .255em;
      vertical-align: .1em;
      content: '';
      border-top: .3em solid;
      border-right: .3em solid transparent;
      border-bottom: 0;
      border-left: .3em solid transparent;
    }
  }

  &__menu {
    position: absolute;
    top: 100%;
    left: auto;
    right: 0;
    bottom: auto;
    z-index: 1000;
    display: none;
    float: left;
    min-width: 10rem;
    padding: .5rem 0;
    margin: .125rem 0 0;
    font-size: 1rem;
    text-align: left;
    list-style: none;
    background-color: #fff;
    background-clip: padding-box;
    border: 1px solid rgba(0,0,0,.15);
    border-radius: .4rem;

    &.show {
      display: block;
    }
  }

  &__item {
    font-family: 'Source Sans Pro', sans-serif;
    display: block;
    width: 100%;
    padding: .25rem 1.5rem;
    clear: both;
    font-size: 1.4rem;
    text-align: inherit;
    white-space: nowrap;
    background-color: transparent;
    border: 0;
    cursor: pointer;

    &:hover {
      background: grey;
      color: currentColor;
    }

    &:focus, &:hover {
      text-decoration: none;
      background-color: #ecedee;
      outline: none;
    }
  }

  &.disabled {

    .dropdown__toggle {
      cursor: default;

      &:after {
        opacity: .5;
      }
    }
  }
}
</style>
