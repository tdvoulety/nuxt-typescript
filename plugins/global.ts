import Vue from 'vue'
import { GLOBAL_COMPONENTS } from '@/constants/global'

GLOBAL_COMPONENTS.forEach((item) => {
  Vue.component(item.key, item.component)
})
