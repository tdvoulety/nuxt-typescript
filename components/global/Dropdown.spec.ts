// @ts-nocheck
import { mountComponent, rendersToSnapshot } from '~/test/unit/helpers'
import Dropdown from '@/components/global/Dropdown'
import { GLOBAL_CLICK_EVENT } from '@/constants/global'

const items = [
  {
    value: 'de',
    text: 'Deutsch',
  },
  {
    value: 'en',
    text: 'English',
  },
  {
    value: 'fr',
    text: 'FranÃ§ais',
  },
]

const mount = (props = {}) => mountComponent(Dropdown, {
  propsData: {
    text: 'text',
    items,
    onClick: jest.fn(),
    ...props,
  },
  stubs: ['router-link'],
})

describe('Dropdown', () => {
  it('shows all items as list items', () => {
    const { wrapper } = mount()

    expect(wrapper.findAll('.dropdown__item').length).toEqual(3)
  })

  it('shows dropdown menu after click on button and hide it after another click on button', async () => {
    const { wrapper } = mount()
    const clickEventOnToggle = { target: { id: wrapper.vm.dropdownId }, stopPropagation: jest.fn() }
    const dropdownMenu = wrapper.find('.dropdown__menu')

    wrapper.vm.$root.$emit(GLOBAL_CLICK_EVENT, clickEventOnToggle)
    await wrapper.vm.$nextTick()
    expect(dropdownMenu.classes('show')).toBe(true)
    wrapper.vm.$root.$emit(GLOBAL_CLICK_EVENT, clickEventOnToggle)
    await wrapper.vm.$nextTick()
    expect(dropdownMenu.classes('show')).toBe(false)
  })

  it('shows dropdown menu after click on button and hide it after click on any element on the page', async () => {
    const { wrapper } = mount()
    const clickEventOnToggle = { target: { id: wrapper.vm.dropdownId }, stopPropagation: jest.fn() }
    const dropdownMenu = wrapper.find('.dropdown__menu')

    wrapper.vm.$root.$emit(GLOBAL_CLICK_EVENT, clickEventOnToggle)
    await wrapper.vm.$nextTick()
    expect(dropdownMenu.classes('show')).toBe(true)
    wrapper.vm.$root.$emit(GLOBAL_CLICK_EVENT)
    await wrapper.vm.$nextTick()
    expect(dropdownMenu.classes('show')).toBe(false)
  })

  it('dropdown has disabled class if prop disabled is true', () => {
    const { wrapper } = mount({ disabled: true })
    const dropdown = wrapper.find('.dropdown')

    expect(dropdown.classes('disabled')).toBe(true)
  })

  it('emit onClick event after click on the item', () => {
    const { wrapper } = mount()

    wrapper.find('.dropdown__item').trigger('click')
    expect(wrapper.emitted().onClick).toBeTruthy()
  })

  it('renders to a snapshot', () => {
    const { wrapper } = mount()

    rendersToSnapshot(wrapper)
  })
})
