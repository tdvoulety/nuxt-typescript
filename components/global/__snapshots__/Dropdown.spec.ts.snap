// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders to a snapshot 1`] = `
<div onClick="function() {
        return fn.apply(this, arguments);
      }" class="dropdown"><button type="button" id="dropdown-25" class="dropdown__toggle"><span id="dropdown-25__text">text</span></button>
  <ul role="menu" tabindex="-1" class="dropdown__menu">
    <li><button type="button" class="dropdown__item">
        Deutsch
      </button></li>
    <li><button type="button" class="dropdown__item">
        English
      </button></li>
    <li><button type="button" class="dropdown__item">
        Fran√ßais
      </button></li>
  </ul>
</div>
`;
